### PR TITLE
cmd/snap: `snap connections` command

### PR DESCRIPTION
**Pull request by bboozzoo**
Early RFC for `snap connections` command, based on the discussion in: https://forum.snapcraft.io/t/rfc-snap-connections-command/4296

Does not show unconnected interfaces by default:
```
$ snap connections test-snapd-service-watchdog
no interface connections found
```

With `-a` shows all interfaces:

```
$ snap connections test-snapd-service-watchdog -a
Plug                                       Slot  Interface      Notes
test-snapd-service-watchdog:daemon-notify  -     daemon-notify  -
```
VLC snap:

```
$ snap connections vlc
Plug                        Slot                     Interface               Notes
vlc:desktop                 :desktop                 desktop                 -
vlc:desktop-legacy          :desktop-legacy          desktop-legacy          -
vlc:home                    :home                    home                    -
vlc:network                 :network                 network                 -
vlc:network-bind            :network-bind            network-bind            -
vlc:opengl                  :opengl                  opengl                  -
vlc:optical-drive           :optical-drive           optical-drive           -
vlc:pulseaudio              :pulseaudio              pulseaudio              -
vlc:removable-media         :removable-media         removable-media         -
vlc:screen-inhibit-control  :screen-inhibit-control  screen-inhibit-control  -
vlc:unity7                  :unity7                  unity7                  -
vlc:x11                     :x11                     x11                     -
$ snap connections vlc -a
Plug                        Slot                     Interface               Notes
vlc:camera                  -                        camera                  -
vlc:desktop                 :desktop                 desktop                 -
vlc:desktop-legacy          :desktop-legacy          desktop-legacy          -
vlc:home                    :home                    home                    -
vlc:mount-observe           -                        mount-observe           -
vlc:network                 :network                 network                 -
vlc:network-bind            :network-bind            network-bind            -
vlc:opengl                  :opengl                  opengl                  -
vlc:optical-drive           :optical-drive           optical-drive           -
vlc:pulseaudio              :pulseaudio              pulseaudio              -
vlc:removable-media         :removable-media         removable-media         -
vlc:screen-inhibit-control  :screen-inhibit-control  screen-inhibit-control  -
vlc:unity7                  :unity7                  unity7                  -
vlc:x11                     :x11                     x11                     -
-                           vlc:mpris                mpris                   -
```

Snap where slot/plug names do not match interface name:

```
$ snap connections gnome-sudoku
Plug                          Slot                             Interface       Notes
gnome-sudoku:desktop          :desktop                         desktop         -
gnome-sudoku:desktop-legacy   :desktop-legacy                  desktop-legacy  -
gnome-sudoku:gnome-3-26-1604  gnome-3-26-1604:gnome-3-26-1604  content         -
gnome-sudoku:gsettings        :gsettings                       gsettings       -
gnome-sudoku:gtk-3-themes     gtk-common-themes:gtk-3-themes   content         -
gnome-sudoku:icon-themes      gtk-common-themes:icon-themes    content         -
gnome-sudoku:pulseaudio       :pulseaudio                      pulseaudio      -
gnome-sudoku:sound-themes     gtk-common-themes:sound-themes   content         -
gnome-sudoku:unity7           :unity7                          unity7          -
gnome-sudoku:wayland          :wayland                         wayland         -
```

System alias:
```
$ snap connections system      
Plug                               Slot                     Interface               Notes
cuda-samples:home                  :home                    home                    -
cuda-samples:network-bind          :network-bind            network-bind            -
cuda-samples:opengl                :opengl                  opengl                  -
cuda-samples:x11                   :x11                     x11                     -
gnome-calculator:desktop           :desktop                 desktop                 -
gnome-calculator:desktop-legacy    :desktop-legacy          desktop-legacy          -
gnome-calculator:gsettings         :gsettings               gsettings               -
gnome-calculator:network           :network                 network                 -
gnome-calculator:unity7            :unity7                  unity7                  -
gnome-calculator:wayland           :wayland                 wayland                 -
gnome-sudoku:desktop               :desktop                 desktop                 -
gnome-sudoku:desktop-legacy        :desktop-legacy          desktop-legacy          -
gnome-sudoku:gsettings             :gsettings               gsettings               -
gnome-sudoku:pulseaudio            :pulseaudio              pulseaudio              -
gnome-sudoku:unity7                :unity7                  unity7                  -
gnome-sudoku:wayland               :wayland                 wayland                 -
http:network                       :network                 network                 -
http:network-bind                  :network-bind            network-bind            -
http:snapd-control                 :snapd-control           snapd-control           -
lxd:lxd-support                    :lxd-support             lxd-support             -
lxd:network                        :network                 network                 -
lxd:network-bind                   :network-bind            network-bind            -
lxd:system-observe                 :system-observe          system-observe          -
ohmygiraffe:network                :network                 network                 -
ohmygiraffe:network-bind           :network-bind            network-bind            -
ohmygiraffe:opengl                 :opengl                  opengl                  -
ohmygiraffe:pulseaudio             :pulseaudio              pulseaudio              -
ohmygiraffe:x11                    :x11                     x11                     -
socket-activation:network-bind     :network-bind            network-bind            -
test-services-25269:daemon-notify  :daemon-notify           daemon-notify           -
vlc:desktop                        :desktop                 desktop                 -
vlc:desktop-legacy                 :desktop-legacy          desktop-legacy          -
vlc:home                           :home                    home                    -
vlc:network                        :network                 network                 -
vlc:network-bind                   :network-bind            network-bind            -
vlc:opengl                         :opengl                  opengl                  -
vlc:optical-drive                  :optical-drive           optical-drive           -
vlc:pulseaudio                     :pulseaudio              pulseaudio              -
vlc:removable-media                :removable-media         removable-media         -
vlc:screen-inhibit-control         :screen-inhibit-control  screen-inhibit-control  -
vlc:unity7                         :unity7                  unity7                  -
vlc:x11                            :x11                     x11                     -
```

### CODE REVIEW

**Comment by zyga** [comment]
It looks very promising. Some random ideas to think about:

 * add --system to show system connections, hide those by default
 * add --gadget or similar for IOT devices, like --system
 * add "manual" note for non-auto connected stuff
 * add "never" note for manually disconnected stuff
 * add "content(foo)" note for content connections 
 * add interface(foo) if plug or slot name doesn't match interface name, remove the interface column 

Some color coding or ordering could be tweaked for usability once we have a better feeling but I think this is super nice already.

File: cmd/snap/cmd_connections.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		}
+	}
+	return wanted == name
+}
+
+func (x *cmdConnections) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+	if x.Positionals.Query.Name != "" {
```
Comment (pedronis) [inline cmd/snap/cmd_connections.go]: why using that type if we don't accept a name in it?

File: cmd/snap/cmd_connections.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		return errors.New("filtering by slot/plug name is not supported")
+	}
+
+	ifaces, err := x.client.Connections()
```
Comment (pedronis) [inline cmd/snap/cmd_connections.go]: ugh, wonder why we called it Connections in client for the current API, we might consider renaming that if we introduce a new one

File: cmd/snap/cmd_connections.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+
+	"github.com/jessevdk/go-flags"
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: nitpick⁹⁹: it' usually system, other, project imports

File: cmd/snap/cmd_connections.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+var shortConnectionsHelp = i18n.G("List interface connections")
+var longConnectionsHelp = i18n.G(`
+List connections between plugs and slots. When passed an optional
+snap name, lists connected and unconnected plugs and slots for
+that snap only.
+
+Pass --all to list connected and unconnected plugs and slots.
+`)
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: long helps should start with "The *foo* command ...".

File: cmd/snap/cmd_connections.go [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+func init() {
+	addCommand("connections", shortConnectionsHelp, longConnectionsHelp, func() flags.Commander {
+		return &cmdConnections{}
+	}, map[string]string{
+		"all":          i18n.G("Show connected and unconnected plugs and slots"),
+		"disconnected": i18n.G("Show disconnected plugs and slots"),
+	}, []argDesc{{
+		// TRANSLATORS: This needs to be wrapped in <>s.
+		name: i18n.G("<snap>"),
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: we don't translate `<snap>`

File: cmd/snap/cmd_connections.go [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+	if isSystemSnap(snap) {
+		return ":" + name
+	}
+	return snap + ":" + name
+}
+
+func wantedSnapMatches(name, wanted string) bool {
+	if wanted == "system" {
+		switch name {
+		case "core", "snapd", "system":
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: why not use `isSystemSnap` here? less duplication that way

File: cmd/snap/cmd_connections.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+	"github.com/jessevdk/go-flags"
+)
+
+type cmdConnections struct {
+	clientMixin
+	All          bool `long:"all"`
+	Disconnected bool `long:"disconnected"`
+	Positionals  struct {
+		Snap installedSnapName `skip-help:"true"`
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: what is `skip-help` supposed to do? (what it actually does is nothing)

File: cmd/snap/cmd_connections.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+	addCommand("connections", shortConnectionsHelp, longConnectionsHelp, func() flags.Commander {
+		return &cmdConnections{}
+	}, map[string]string{
+		"all":          i18n.G("Show connected and unconnected plugs and slots"),
+		"disconnected": i18n.G("Show disconnected plugs and slots"),
+	}, []argDesc{{
+		// TRANSLATORS: This needs to be wrapped in <>s.
+		name: i18n.G("<snap>"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		desc: i18n.G("Constrain listing to a specific snap"),
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: I think descriptions should be full sentences, but I know they currently aren't, everywhere. We're inconsistent. We should discuss it and fix it sometime.

File: cmd/snap/cmd_connections.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+		// only
+		x.All = !x.Disconnected
+	}
+
+	connections, err := x.client.Connections(&opts)
+	if err != nil {
+		return err
+	}
+	if len(connections.Plugs) == 0 && len(connections.Slots) == 0 {
+		return fmt.Errorf(i18n.G("no connections found"))
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: It isn't an error if a snap has no connections.

File: cmd/snap/cmd_connections.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+			plug:      endpoint(conn.Plug.Snap, conn.Plug.Name),
+			slot:      endpoint(conn.Slot.Snap, conn.Slot.Name),
+			manual:    conn.Manual,
+			gadget:    conn.Gadget,
+			undesired: true,
+		}
+	}
+
+	w := tabWriter()
+	fmt.Fprintln(w, i18n.G("Plug\tSlot\tInterface\tNotes"))
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: If a snap only has connected and the user asked for disconnected, we'll print the header and then nothing more.

File: cmd/snap/cmd_connections.go [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+	opts := client.ConnectionOptions{
+		All: x.All || x.Disconnected,
+	}
+	wanted := string(x.Positionals.Snap)
+	if wanted != "" {
+		// when asking for a single snap, include its disconnected plugs
+		// and slots
+		opts.Snap = wanted
+		opts.All = true
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: Given this behaviour, I'd say error out if `--all` is given together with a snap name. Otherwise it's confusing.

**Comment by chipaca** [comment]
FWIW connections might benefit from dimming repeated bits in its listing:
image
(if we want to do #6482)

File: cmd/snap/cmd_connections.go (hunk 10 above)
Comment (bboozzoo) [reply cmd/snap/cmd_connections.go]: Fixed

File: cmd/snap/cmd_connections.go (hunk 4 above)
Comment (bboozzoo) [reply cmd/snap/cmd_connections.go]: Tried to tweak the help information a little, but it's likely I came up with something even worse in the process. Would appreciate your and @degville's input on this.

Comment (degville) [reply cmd/snap/cmd_connections.go]: I think it's better to be consistent with our other long help output (eg. aliases or interfaces), which would make it similar to the following:

```
The connections command lists connections between plugs and slots
in the system. Pass --all to list unconnected plugs and slots 
alongside those already connected.

$ snap connections <snap>

Lists connected and unconnected plugs and slots for the specified
snap.                                                         
```

**Comment by pedronis** [comment]
We need to think a bit more at least what goes into the Notes column,  using "disconnected "for both  not-yet connected slots or plugs, or  a connection that was  "snap disconnect"-ed feels too fuzzy/slightly confusing

File: cmd/snap/cmd_connections.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func endpoint(snap, name string) string {
+	if isSystemSnap(snap) {
+		return ":" + name
+	}
+	return snap + ":" + name
+}
+
+func wantedSnapMatches(name, wanted string) bool {
```
Comment (pedronis) [inline cmd/snap/cmd_connections.go]: this doesn't seem to be used, is it needed? missing tests?

File: cmd/snap/cmd_connections.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+			notes = append(notes, connectionNotes{
+				plug:          endpoint(plug.Snap, plug.Name),
+				slot:          "-",
+				interfaceName: plug.Interface,
+			})
+		}
+	}
+	for _, slot := range connections.Slots {
+		if isSystemSnap(slot.Snap) {
+			// displaying unconnected system snap slots is boring
```
Comment (pedronis) [inline cmd/snap/cmd_connections.go]: mmh, should we do it though for an explicit "snap connections system" ?

File: cmd/snap/cmd_connections.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+	if wanted == "system" {
+		if isSystemSnap(name) {
+			return true
+		}
+		return false
+	}
+	return wanted == name
+}
+
+type connectionNotes struct {
```
Comment (pedronis) [inline cmd/snap/cmd_connections.go]: s/Notes// ?   this more than just the notes

File: cmd/snap/cmd_help.go [hunk 15]
```diff
@@ -198,7 +198,7 @@ var helpCategories = []helpCategory{
 	}, {
 		Label:       i18n.G("Permissions"),
 		Description: i18n.G("manage permissions"),
-		Commands:    []string{"interfaces", "interface", "connect", "disconnect"},
+		Commands:    []string{"connections", "interfaces", "interface", "connect", "disconnect"},
```
Comment (pedronis) [inline cmd/snap/cmd_help.go]: should we stop listing "interfaces" here?  wait for 2.39 ?  cc @chipaca

File: cmd/snap/cmd_connections.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+var shortConnectionsHelp = i18n.G("List interface connections")
+var longConnectionsHelp = i18n.G(`
+The connections command lists connections between plugs and slots
+in the system.
+
+Pass --all to list unconnected plugs and slots alongside those
+already connected. Unless <snap> is provided, the listing is for all
+snaps in the system.
```
Comment (chipaca) [inline cmd/snap/cmd_connections.go]: I think we should flip this paragraph, and add a little bit to make the behaviour clearer, something like
```
Unless <snap> is provided, the listing is for connected plugs and slots
for all snaps in the system.  In this mode, pass --all to also list
unconnected plugs and slots.
```

File: cmd/snap/cmd_connections.go (hunk 12 above)
Comment (chipaca) [reply cmd/snap/cmd_connections.go]: `unused` (from `go get honnef.co/go/tools/cmd/unused`) says no:
```
$ unused ./cmd/snap
Unused has been deprecated. Please use staticcheck instead.
cmd/snap/cmd_advise.go:114:3: field Command is unused (U1000)
cmd/snap/cmd_advise.go:115:3: field SearchTerms is unused (U1000)
cmd/snap/cmd_connections.go:80:6: func wantedSnapMatches is unused (U1000)
cmd/snap/cmd_connections.go:112:6: func connName is unused (U1000)
cmd/snap/cmd_connectivity_check.go:47:3: field Connectivity is unused (U1000)
cmd/snap/cmd_snapshot.go:33:6: func fmtSize is unused (U1000)
cmd/snap/cmd_snapshot_test.go:32:6: type snapshotCmdArgs is unused (U1000)
```

File: cmd/snap/cmd_connections.go [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+	if cn.gadget {
+		opts = append(opts, "gadget")
+	}
+	if len(opts) == 0 {
+		return "-"
+	}
+	return strings.Join(opts, ",")
+}
+
+func connName(conn client.Connection) string {
```
[the remark this answers is not in the document]
Comment (chipaca) [reply cmd/snap/cmd_connections.go]: as above, confirm not used

File: cmd/snap/cmd_help.go (hunk 15 above)
Comment (chipaca) [reply cmd/snap/cmd_help.go]: I think we should drop it (it's confusing, and this is better). Further, I think `snap interfaces` should print a 'try the new connections' to stderr.

**Comment by degville** [comment]
> One thing I found surprising was that `snap connections some-thing-that-does-not-exist` would not error. Is this on purpose? _Can_ we detect a "snap does not exist" versus a "snap has no connections"? (maybe if there are no connections, query whether the snap exists?)

As an aside to this, I filed a bug with `snap interfaces` for a similar issue:
https://bugs.launchpad.net/snapd/+bug/1816735

**Comment by pedronis** [comment]
I think it would be fine for the new API to return a snap-not-found error if the snap filter is given and the snap does not exist

File: cmd/snap/cmd_connections.go [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+
+	connections, err := x.client.Connections(&opts)
+	if err != nil {
+		return err
+	}
+	if len(connections.Plugs) == 0 && len(connections.Slots) == 0 {
+		return nil
+	}
+
+	notes := make([]connection, 0, len(connections.Established)+len(connections.Undesired))
```
Comment (pedronis) [inline cmd/snap/cmd_connections.go]: we might want a different name here now,  s/notes/conns/ ?   s/notes/annotatedConns/ ?

**Comment by bboozzoo** [comment]
Yay!

**Pull request merged**